wakuv2/waku.go

```diff
@@ -514,8 +514,13 @@ func (w *Waku) retryDnsDiscoveryWithBackoff(ctx context.Context, addr string, su
 		}
 
 		t := time.NewTimer(backoff)
+		c := w.ctx
+		// c might be nil when w.Stop is invoked while this goroutine is running
+		if c == nil {
+			return
+		}
 		select {
-		case <-w.ctx.Done():
+		case <-c.Done():
```

Comment:
So there still might be a race condition, it's just less probable. The thread could be switched after the `c == nil` check, but before you call `<-c.Done()`. 

I guess it would still work even in this case, because you actually copy the `w.ctx`.
But it's quite a nasty fix I think.

I think we should just add `w.wg.Add(1)` and `w.wg.Done()` to the goroutine here:
https://github.com/status-im/status-go/blob/a4ca18a14f5628f542a70213cfe33baece12ffa2/wakuv2/waku.go#L423-L426

Then `Waku.Stop()` will wait till all goroutines are finihsed:
https://github.com/status-im/status-go/blob/a4ca18a14f5628f542a70213cfe33baece12ffa2/wakuv2/waku.go#L1457

Reply:
@qfrank in your example this is wrong:
```go
	t.Log("Main: Setting ctx and cancel to nil")
	s.ctx = nil
	s.cancel = nil

	t.Log("Main: Waiting for goroutines to complete")
	s.wg.Wait()
```

We should first `wg.Wait()` and only then `s.ctx = nil`.
This is exactly how we do in `Stop()`:
https://github.com/status-im/status-go/blob/a4ca18a14f5628f542a70213cfe33baece12ffa2/wakuv2/waku.go#L1457-L1460

Reply:
> but got no response from status-go-guild on discord

Pardon, busy times lately. 
Sure any refactoring is supported! 👍 Feel free to take anything, just coordinate your actions with guild members and @plopezlpz before starting 🙂 

> I did it intentionally, otherwise we won't see "Context is nil, exiting safely"

Not sure what you mean then. I thought you're showing this code as an example that this won't work.

What I'm saying here, that instead of applying a workaround that you're making in this PR, we should just use the `waku.wg` here, same as it is done for all other goroutines in this module.